python/taichi/lang/ast/ast_transformer.py

```diff
@@ -582,8 +582,58 @@ def transform_as_kernel():
                     # Template arguments are passed by reference.
                     if isinstance(ctx.func.arguments[i].annotation,
                                   annotations.template):
+
+                        ctx.create_variable(ctx.func.arguments[i].name, data)
+                        continue
+
+                    # Ndarray arguments are passed by reference.
```

Comment:
This part should behave the same as https://github.com/taichi-dev/taichi/blob/365e755190643c9f4504fbe7f990bd167762dfe0/python/taichi/lang/kernel_impl.py#L345-L355

Reply:
hmmm, AnyArray doesn't have "check_matched()" and "get_type()" yet, I'll leave dtype check for AnyArray as is for now

Reply:
`check_matched()` is already there. You need to implement `get_type()` though.

Reply:
done, thanks!